cmd/ocsp-updater/main.go

```diff
@@ -212,49 +414,64 @@ func main() {
 		blog.SetAuditLogger(auditlogger)
 
 		go cmd.DebugServer(c.OCSPUpdater.DebugAddr)
+		go cmd.ProfileCmd("OCSP-Updater", stats)
 
 		// Configure DB
 		dbMap, err := sa.NewDbMap(c.OCSPUpdater.DBConnect)
 		cmd.FailOnError(err, "Could not connect to database")
 
-		cac, closeChan := setupClients(c, stats)
+		cac, pubc, closeChan := setupClients(c, stats)
+
+		updater, err := newUpdater(
+			auditlogger,
+			stats,
+			clock.Default(),
+			dbMap,
+			cac,
+			pubc,
+			// Necessary evil for now
+			len(c.Publisher.CT.Logs),
+			c.OCSPUpdater,
+		)
+		cmd.FailOnError(err, "Failed to create updater")
 
+		auditlogger.Info(app.VersionString())
+
+		// Handle AMQP disconnections gracefully (and block other operations while
+		// disconnected...)
 		go func() {
-			// Abort if we disconnect from AMQP
 			for {
-				for err := range closeChan {
-					auditlogger.Warning(fmt.Sprintf(" [!] AMQP Channel closed, aborting early: [%s]", err))
-					panic(err)
-				}
+				err := <-closeChan
+				auditlogger.Warning(fmt.Sprintf(" [!] AMQP Channel closed, will reconnect in 5 seconds: [%s]", err))
+				time.Sleep(5 * time.Second)
+				updater.rpcMu.Lock()
+				updater.cac, updater.pub, closeChan = setupClients(c, stats)
+				updater.rpcMu.Unlock()
+				auditlogger.Info(" [!] Reconnected to AMQP Channel")
 			}
 		}()
 
-		auditlogger.Info(app.VersionString())
-
-		updater := &OCSPUpdater{
-			cac:   cac,
-			dbMap: dbMap,
-			stats: stats,
-			log:   auditlogger,
-		}
-
-		// Calculate the cut-off timestamp
-		if c.OCSPUpdater.MinTimeToExpiry == "" {
-			panic("Config must specify a MinTimeToExpiry period.")
-		}
-		dur, err := time.ParseDuration(c.OCSPUpdater.MinTimeToExpiry)
-		cmd.FailOnError(err, "Could not parse MinTimeToExpiry from config.")
-
-		oldestLastUpdatedTime := time.Now().Add(-dur)
-		auditlogger.Info(fmt.Sprintf("Searching for OCSP responses older than %s", oldestLastUpdatedTime))
-
-		// When we choose to batch responses, it may be best to restrict count here,
-		// change the transaction to survive the whole findStaleResponses, and to
-		// loop this method call however many times is appropriate.
-		err = updater.findStaleResponses(oldestLastUpdatedTime, c.OCSPUpdater.ResponseLimit)
-		if err != nil {
-			auditlogger.WarningErr(err)
-		}
+		go updater.newCertLoop()
+		go updater.oldOCSPLoop()
+		go updater.missingSCTLoop()
```

Comment:
yeah, each of these loops implies at least one different struct to me.

I've not sat down and worked out which things share dependencies but I assume you can find a nice place to cut.

Also, "oldOCSPLoop" and "newCertLoop" aren't documented as to why they are called "old" and "new". What's up with them?